codalab/rest/groups.py

```diff
@@ -0,0 +1,217 @@
+    local.model.update_group(patch)
+    group.update(patch)
+    return GroupSchema().dump(group).data
+
+
+@post('/groups', apply=AuthenticatedPlugin())
+def create_group():
+    """Create a group."""
+    try:
+        group = GroupSchema(strict=True).load(request.json, partial=True).data
+    except ValidationError as err:
+        abort(httplib.BAD_REQUEST, err.messages)
+    ensure_unused_group_name(group['name'])
+    group['owner_id'] = request.user.user_id
+    group = local.model.create_group(group)
+    local.model.add_user_in_group(request.user.user_id, group['uuid'], True)
+    return GroupSchema().dump(group).data
+
+
+#TODO(sckoo): include role as meta!!
+@get('/groups', apply=AuthenticatedPlugin())
+def fetch_groups():
+    """Fetch list of groups readable by the authenticated user."""
+    if request.user.user_id == local.model.root_user_id:
+        groups = local.model.batch_get_all_groups(
+            None,
+            {'user_defined': True},
+            None
+        )
+    else:
+        groups = local.model.batch_get_all_groups(
+            None,
+            {'owner_id': request.user.user_id, 'user_defined': True},
+            {'user_id': request.user.user_id}
+        )
+
+    user_ids = set()
+    memberships = []
+    for group in groups:
+        group['memberships'] = local.model.batch_get_user_in_group(group_uuid=group['uuid'])
+        memberships.extend(group['memberships'])
+        user_ids |= set(m['user_id'] for m in group['memberships'])
+
+    document = GroupSchema(many=True).dump(groups).data
+    document['included'] = []
+    document['included'].extend(
+        GroupMembershipSchema(many=True).dump(memberships).data['data']
+    )
+    document['included'].extend(
+        UserSchema(many=True, only=PUBLIC_USER_FIELDS).dump(local.model.get_users(user_ids)).data['data']
+    )
+
+    return document
+
+
+@post('/group-memberships', apply=AuthenticatedPlugin())
+def create_or_update_group_membership():
+    try:
+        membership_updates = GroupMembershipSchema(strict=True).load(request.json).data
+    except ValidationError as err:
+        abort(httplib.BAD_REQUEST, err.messages)
+
+    # FIXME(sckoo): This may create race condition, should just combine these bundle model methods
+    memberships = local.model.batch_get_user_in_group(user_id=membership_updates['user_id'], group_uuid=membership_updates['group_uuid'])
+    if len(memberships) > 0:
+        membership = memberships[0]
+        membership.update(membership_updates)
+        local.model.update_user_in_group(membership['user_id'], membership['group_uuid'], membership['is_admin'])
+    else:
+        membership = membership_updates
+        membership = local.model.add_user_in_group(membership['user_id'], membership['group_uuid'], membership['is_admin'])
+
+    return GroupMembershipSchema().dump(membership).data
+
+
+@delete('/group-memberships/<membership_id>', apply=AuthenticatedPlugin())
+def delete_group_membership(membership_id):
```

Comment:
Is delete with a single membership_id standard?

I thought that with the JSON API, you would be sending some sort of JSON message to delete.

What if you wanted to delete multiple memberships in batch?

Don't you also need to check that the user has permissions to delete this?